pkg/k8shandler/clusterrestart_test.go

```diff
@@ -0,0 +1,445 @@
+			expectedStatus.UpgradeStatus.UpgradePhase = api.PreparationComplete
+
+			Expect(restarter.restartCluster()).NotTo(BeNil())
+			Expect(restarter.nodeStatus).To(BeEquivalentTo(expectedStatus))
+		})
+	})
+
+	Context("node fails post", func() {
+		JustBeforeEach(func() {
+
+			testNode := &deploymentNode{}
+			testNodes = []NodeTypeInterface{
+				testNode,
+			}
+
+			restarter = Restarter{
+				scheduledNodes:   testNodes,
+				clusterName:      "test-cluster",
+				clusterNamespace: "test-namespace",
+				precheck:         r.restartNoop,
+				prep:             r.restartNoop,
+				main:             r.restartNoop,
+				post:             r.restartFail,
+			}
+
+			restarter.nodeStatus = nodeStatus
+			restarter.setNodeConditions(updateStatus)
+		})
+
+		It("should fail to complete restarting", func() {
+			expectedStatus := nodeStatus
+			expectedStatus.UpgradeStatus.UnderUpgrade = v1.ConditionTrue
+			expectedStatus.UpgradeStatus.ScheduledForUpgrade = ""
+			expectedStatus.UpgradeStatus.UpgradePhase = api.NodeRestarting
+
+			Expect(restarter.restartCluster()).NotTo(BeNil())
+			Expect(restarter.nodeStatus).To(BeEquivalentTo(expectedStatus))
+		})
+	})
+
+	Context("node fails recovery", func() {
+		JustBeforeEach(func() {
+
+			testNode := &deploymentNode{}
+			testNodes = []NodeTypeInterface{
+				testNode,
+			}
+
+			restarter = Restarter{
+				scheduledNodes:   testNodes,
+				clusterName:      "test-cluster",
+				clusterNamespace: "test-namespace",
+				precheck:         r.restartNoop,
+				prep:             r.restartNoop,
+				main:             r.restartNoop,
+				post:             r.restartNoop,
+				recovery:         r.restartFail,
+			}
+
+			restarter.nodeStatus = nodeStatus
+			restarter.setNodeConditions(updateStatus)
+		})
+
+		It("should fail to complete restarting", func() {
+			expectedStatus := nodeStatus
+			expectedStatus.UpgradeStatus.UnderUpgrade = v1.ConditionTrue
+			expectedStatus.UpgradeStatus.ScheduledForUpgrade = ""
+			expectedStatus.UpgradeStatus.UpgradePhase = api.RecoveringData
+
+			Expect(restarter.restartCluster()).NotTo(BeNil())
+			Expect(restarter.nodeStatus).To(BeEquivalentTo(expectedStatus))
+		})
+	})
+
+	// ---------------------
+	// cluster restart tests
+	// ---------------------
+
+	// test to ensure the cluster restart gating progresses all the way through
+	Context("cluster restarts", func() {
+		JustBeforeEach(func() {
+			restarter = Restarter{
+				scheduledNodes:   testNodes,
+				clusterName:      "test-cluster",
+				clusterNamespace: "test-namespace",
+				precheck:         r.restartNoop,
+				prep:             r.restartNoop,
+				main:             r.restartNoop,
+				post:             r.restartNoop,
+				recovery:         r.restartNoop,
+			}
+
+			restarter.clusterStatus = clusterStatus
+			restarter.setClusterConditions(updateStatus)
+		})
+
+		It("should be able to complete without error", func() {
+			expectedStatus := clusterStatus
```

Comment:
Why is a successful cluster restart having the same expected status as `cluster fails precheck` and `node fails precheck`? Is this maybe a missing status condition?

Reply:
no, it is the nature of how we gate. being completed is the same as never progressing past our first step since we would need to retry that first step -- we need to be able to pass the precheck condition again.